src/browser/components/TeamList.jsx

```diff
@@ -136,23 +135,43 @@ export default class TeamList extends React.Component {
             url: newTeam.url,
             order: newTeam.order,
           };
-          if (this.props.showAddTeamForm) {
-            this.props.addServer(teamData);
-          } else {
-            this.props.updateTeam(newTeam.index, teamData);
-          }
-          this.setState({
-            showNewTeamModal: false,
-            showEditTeamForm: false,
-            team: {
-              name: '',
-              url: '',
-              index: false,
-              order: newTeam.order + 1,
-            },
+
+          // Check to ensure url is a Mattermost server
+          fetch(newTeam.url).then((response) => {
+            return response.text();
+          }).then((html) => {
+            const parser = new DOMParser();
+            const doc = parser.parseFromString(html, 'text/html');
+            let titleTag = doc.getElementsByTagName('title');
+            if (titleTag[0]) {
+              titleTag = titleTag[0];
+            }
+            if (titleTag && titleTag.innerHTML === 'Mattermost') {
+              // It is a Mattermost server, so proceed with adding it
+              if (this.props.showAddTeamForm) {
+                this.props.addServer(teamData);
+              } else {
+                this.props.updateTeam(newTeam.index, teamData);
+              }
+              this.setState({
+                showNewTeamModal: false,
+                showEditTeamForm: false,
+                team: {
+                  name: '',
+                  url: '',
+                  index: false,
+                  order: newTeam.order + 1,
+                },
+              });
+              this.render();
+              this.props.setAddTeamFormVisibility(false);
+            } else {
+              alert('This does not appear to be a Mattermost server.'); // eslint-disable-line no-alert
+            }
+          }).catch((err) => {
```

Comment:
I agree.

To address this and the comment below, it seems the best way to refactor this would be to ditch the native alert and go with a custom UI so we can design the modal how we like.

We'd also ideally want the modal to pop up instantly and show a loading spinner with a cancel button to address the network slowness use case.

For the error case we could offer a "force add" button that would let the user add the server without this safety check, but maybe put up a big scary warning that you shouldn't do it unless you know what you're doing.

I'm not sure how to approach building though though. Any advice?

Reply:
definitely! The popup should show as soon as possible.

You can find some examples of modals in the code already, one of the most recents is [the permission modal](https://github.com/mattermost/desktop/blob/master/src/browser/components/PermissionModal.jsx)